File: section-06-basic-strings/06-rotate-the-string/rotate-the-string-opt.cs
using System;

public class Solution
{
    // Strings must be of the same length to be rotations of each other
    public bool rotateString(string s, string goal)
    {
        if (s.Length != goal.Length)
        {
            return false;
        }
        string doubledS = s + s;  // Concatenate s with itself
        return doubledS.Contains(goal);  // Check if goal is a substring of s + s
    }
}

/* 

Given two strings s and goal, return true if and only if s can become goal after some number of shifts on s.
A shift on s consists of moving the leftmost character of s to the rightmost position.
For example, if s = "abcde", then it will be "bcdea" after one shift.

Input : s = "abcde" , goal = "cdeab"

Output : true

Explanation :

After performing 2 shifts we can achieve the goal string from string s.
After first shift the string s is => bcdea
After second shift the string s is => cdeab.

*/

File: section-06-basic-strings/06-rotate-the-string/rotate-the-string-opt.py
class Solution:
# Strings must be of the same length to be rotations of each other
    def rotateString(self, s: str, goal: str) -> bool:
        if len(s) != len(goal):
            return False  
        doubled_s = s + s  # Concatenate s with itself
        return goal in doubled_s  # Check if goal is a substring of s + s

# Test cases
sol = Solution()
print(sol.rotateString("abcde", "cdeab")) 
print(sol.rotateString("abcde", "abced")) 



""" 
Time Complexity O(N) , because checking for a substring in s + s is linear in time.

Space Complexity O(N) for the space needed to store the concatenated string s + s.

"""